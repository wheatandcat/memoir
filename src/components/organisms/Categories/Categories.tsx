import React, { memo } from 'react';
import { StyleSheet } from 'react-native';
import View from 'components/atoms/View';
import CategoryButton from 'components/molecules/CategoryButton';

type Props = {
  categoryID: number | null;
  onPress: (categoryID: number) => void;
};

const categoryItems = [
  {
    id: 1,
    name: '趣味',
  },
  {
    id: 2,
    name: '家事',
  },
  {
    id: 3,
    name: '仕事',
  },
  {
    id: 4,
    name: '趣味',
  },
  {
    id: 5,
    name: '家事',
  },
  {
    id: 6,
    name: '仕事',
  },
  {
    id: 7,
    name: '趣味',
  },
  {
    id: 8,
    name: '家事',
  },
  {
    id: 9,
    name: '仕事',
  },
];

const Categories: React.FC<Props> = (props) => {
  return (
    <View style={styles.root}>
      {categoryItems.map((v) => (
        <View key={v.id} my={1}>
          <CategoryButton
            selected={v.id === props.categoryID}
            name={v.name}
            onPress={() => props.onPress(v.id)}
          />
        </View>
      ))}
    </View>
  );
};

export default memo(Categories);

const styles = StyleSheet.create({
  root: {
    flexDirection: 'row',
    flexWrap: 'wrap',
    justifyContent: 'space-between',
    alignItems: 'center',
  },
});
