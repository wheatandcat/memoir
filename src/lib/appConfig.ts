import firebase from 'lib/system/firebase';

export type AppConfig = {
  maintenance: boolean;
  maintenanceMessage: string;
  maintenancePeriod: string | null;
  supportVersion: string;
};

export const defaultAppConfig = () => ({
  maintenance: false,
  maintenanceMessage: '',
  maintenancePeriod: null,
  supportVersion: '1.0.0',
});

export const getAppConfig = async (
  db: firebase.firestore.Firestore
): Promise<AppConfig> => {
  const qs = await db.collection('appConfig').limit(1).get();

  const records = qs.docs.map((elem) => {
    return elem.data();
  });

  if (!records || !records[0]) {
    return defaultAppConfig();
  }

  return records[0] as AppConfig;
};
