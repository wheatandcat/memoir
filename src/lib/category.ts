import { StyleSheet, ViewStyle } from 'react-native';
import theme from 'config/theme';
import master from 'lib/master';

export const categoryBorderStyle = (category: number): ViewStyle => {
  switch (category) {
    case master.CATEGORY_1:
      return styles.category1;
    case master.CATEGORY_2:
      return styles.category2;
    case master.CATEGORY_3:
      return styles.category3;
  }

  return styles.nonCategory;
};

const styles = StyleSheet.create({
  nonCategory: {
    borderLeftColor: theme().color.background.main,
  },
  category1: {
    borderLeftColor: theme().category.color.category1,
  },
  category2: {
    borderLeftColor: theme().category.color.category4,
  },
  category3: {
    borderLeftColor: theme().category.color.category2,
  },
});
