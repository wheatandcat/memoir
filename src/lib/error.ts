const CodeDefault = '-1' as const;
// バリデーションエラー
const CodeValidation = '000001' as const;
// 無効な認証
const CodeInvalidAuthorization = '000002' as const;
// Not Found
const CodeNotFound = '000003' as const;
// Already Exists
const CodeAlreadyExists = '000004' as const;
// 自身の招待コード
const CodeMyInviteCode = '000005' as const;

export const errorCode = {
  CodeDefault,
  CodeValidation,
  CodeInvalidAuthorization,
  CodeNotFound,
  CodeAlreadyExists,
  CodeMyInviteCode,
};
