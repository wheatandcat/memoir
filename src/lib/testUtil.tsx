import React from 'react';
import { GraphQLHandler, GraphQLRequest } from 'msw';
import {
  ApolloClient,
  ApolloProvider,
  createHttpLink,
  InMemoryCache,
} from '@apollo/client';
import { render } from '@testing-library/react-native';
import { server } from 'mocks/server';
import fetch from 'cross-fetch';

const link = createHttpLink({
  uri: 'http://localhost:8080/query',
  fetch,
  credentials: 'same-origin',
});

const client = new ApolloClient({
  link,
  uri: 'http://localhost:8080/query',
  cache: new InMemoryCache(),
});

export const testRenderer =
  (children: React.ReactNode) =>
  (responseOverride?: GraphQLHandler<GraphQLRequest<never>>) => {
    if (responseOverride) {
      server.use(responseOverride);
    }
    return render(<ApolloProvider client={client}>{children}</ApolloProvider>);
  };
