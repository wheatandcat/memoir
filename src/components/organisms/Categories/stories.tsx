import React from 'react';
import { StyleSheet } from 'react-native';
import { storiesOf } from '@storybook/react-native';
import { mockFn } from 'storyBookUtils/index';
import View from 'components/atoms/View';
import Categories from './Categories';

storiesOf('organisms', module).add('Categories', () => (
  <View style={styles.root}>
    <View style={styles.inner}>
      <Categories categoryID={null} onPress={mockFn('onPress')} />
    </View>
  </View>
));

const styles = StyleSheet.create({
  root: {
    width: '100%',
  },
  inner: {
    justifyContent: 'center',
    alignItems: 'center',
    width: 300,
  },
});
