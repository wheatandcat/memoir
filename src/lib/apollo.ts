import { setContext } from '@apollo/client/link/context';
import { Alert } from 'react-native';
import { onError } from '@apollo/client/link/error';
import { ApolloClient, InMemoryCache, createHttpLink } from '@apollo/client';
import { storageKey, getItem } from 'lib/storage';
import Auth from 'lib/auth';

export const cache = new InMemoryCache();
const auth = new Auth();

type Param = {
  Authorization?: string;
  UserID?: string;
};

const makeApolloClient = async () => {
  const uri = `${process.env.API_HOST}/query`;

  const authLink = setContext(async (_, { headers }) => {
    const h = headers;
    const param: Param = {};
    const token = await auth.getIdToken();

    if (token) {
      param.Authorization = `Bearer ${token}`;
    } else {
      const userID = await getItem(storageKey.USER_ID_KEY);

      if (userID) {
        param.UserID = userID;
      }
    }

    return {
      headers: {
        ...h,
        ...param,
      },
    };
  });

  const errorLink = onError((error) => {
    console.log(error.graphQLErrors);
    if ((error.graphQLErrors || []).length > 0) {
      const graphQLErrors = error.graphQLErrors || [
        { message: 'エラー発生しました' },
      ];

      const message = graphQLErrors[0].message;
      Alert.alert('エラー', message);
    }
  });

  return new ApolloClient({
    link: errorLink.concat(authLink.concat(createHttpLink({ uri }))),
    cache: cache,
  });
};

export default makeApolloClient;
