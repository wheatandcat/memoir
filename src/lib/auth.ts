import firebase from 'lib/system/firebase';
import AsyncStorage from '@react-native-async-storage/async-storage';
import { storageKey } from 'lib/storage';
import dayjs from 'lib/dayjs';

const initFirebaseAuth = (): Promise<firebase.User | null> => {
  return new Promise((resolve) => {
    var unsubscribe = firebase.auth().onAuthStateChanged((user) => {
      // user オブジェクトを resolve
      resolve(user);

      // 登録解除
      unsubscribe();
    });
  });
};

class Auth {
  setSession = async (refresh = false) => {
    const user = await initFirebaseAuth();
    if (!user) {
      return null;
    }

    await AsyncStorage.setItem(storageKey.AUTH_UID_KEY, user.uid);

    const result = await user.getIdTokenResult(refresh);

    await AsyncStorage.setItem(storageKey.AUTH_ID_TOKEN_KEY, result.token);
    await AsyncStorage.setItem(
      storageKey.AUTH_ID_TOKEN_EXPIRATION_KEY,
      String(result.claims.exp)
    );

    return result.token;
  };
  getIdToken = async () => {
    const idToken = await AsyncStorage.getItem(storageKey.AUTH_ID_TOKEN_KEY);
    if (!idToken) {
      return null;
    }

    const expiration = await AsyncStorage.getItem(
      storageKey.AUTH_ID_TOKEN_EXPIRATION_KEY
    );

    if (Number(expiration) > dayjs().unix()) {
      return idToken;
    }

    return this.setSession(true);
  };
  logout = async () => {
    await firebase.auth().signOut();

    await AsyncStorage.removeItem(storageKey.AUTH_UID_KEY);
    await AsyncStorage.removeItem(storageKey.AUTH_ID_TOKEN_KEY);
    await AsyncStorage.removeItem(storageKey.AUTH_ID_TOKEN_EXPIRATION_KEY);
  };
}

export default Auth;
