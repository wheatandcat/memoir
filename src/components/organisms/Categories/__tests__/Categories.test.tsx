import React from 'react';
import { testRenderer } from 'lib/testUtil';
import { screen } from '@testing-library/react-native';
import Categories, { Props } from '../';

const propsData = (): Props => ({
  categoryID: 1,
  onPress: jest.fn(),
});

describe('components/organisms/Categories/Categories.tsx', () => {
  it('正常にrenderすること', () => {
    testRenderer(<Categories {...propsData()} />)();
    expect(screen.findAllByText('特別なこと')).toBeTruthy();
  });
});
