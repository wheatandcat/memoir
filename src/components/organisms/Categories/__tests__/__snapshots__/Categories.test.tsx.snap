// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/organisms/Categories/Categories.tsx 正常にrenderすること 1`] = `
<Fragment>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <View
      key="1"
      my={1}
    >
      <Memo(CategoryButton)
        categoryID={1}
        onPress={[Function]}
        selected={true}
      />
    </View>
    <View
      key="10"
      my={1}
    >
      <Memo(CategoryButton)
        categoryID={10}
        onPress={[Function]}
        selected={false}
      />
    </View>
    <View
      key="4"
      my={1}
    >
      <Memo(CategoryButton)
        categoryID={4}
        onPress={[Function]}
        selected={false}
      />
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <View
      key="3"
      my={1}
    >
      <Memo(CategoryButton)
        categoryID={3}
        onPress={[Function]}
        selected={false}
      />
    </View>
    <View
      key="2"
      my={1}
    >
      <Memo(CategoryButton)
        categoryID={2}
        onPress={[Function]}
        selected={false}
      />
    </View>
    <View
      key="5"
      my={1}
    >
      <Memo(CategoryButton)
        categoryID={5}
        onPress={[Function]}
        selected={false}
      />
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <View
      key="13"
      my={1}
    >
      <Memo(CategoryButton)
        categoryID={13}
        onPress={[Function]}
        selected={false}
      />
    </View>
    <View
      key="6"
      my={1}
    >
      <Memo(CategoryButton)
        categoryID={6}
        onPress={[Function]}
        selected={false}
      />
    </View>
    <View
      key="12"
      my={1}
    >
      <Memo(CategoryButton)
        categoryID={12}
        onPress={[Function]}
        selected={false}
      />
    </View>
  </View>
</Fragment>
`;
